### ISSUE TITLE
Casting functions with different context types should produce error

### ISSUE THREAD

**Issue by tomblind** [Perryvw/TypescriptToLua#309]
From https://github.com/Perryvw/TypescriptToLua/pull/307#issuecomment-448616990

In the new functions branch, casting functions with context to ones without and vice versa should produce an error:
```ts
interface Callbacks { callback:(this:void,i:number)=>number }
let callbacks = <Callbacks>{ callback:(i)=>{ return 2*i; } } // Should error
callbacks.callback(2) // arithmetic on nil
```

**Comment by Istar-Eldritch** [comment]
It's easy to understand the reasoning, but I do not know if it would be easy to actually deal with this. Typescript variance system dictates that in cases like this the callback accepting fewer arguments is a valid type, the `i` argument would be inferred as `this` and the other argument would be ignored.
https://basarat.gitbooks.io/typescript/docs/types/type-compatibility.html#variance

How where you intending on solving this issue @tomblind?

**Comment by tomblind** [comment]
This isn't actually a case of different parameter counts, since every function has a 'this' parameter. This is a case of differing types on that parameter. TS normally allows this since the default type of 'this' is 'any' so we need to catch and throw our own error explicitly. With the new function stuff, this check already happens when an object is assigned to variable with an explicit type, so I'm thinking we should be able to run that exact same check when casting as well.
